### PR TITLE
Multithreading

### PR DESCRIPTION
**Pull request by afilini**

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -250,21 +262,119 @@ impl<S: Read + Write> Client<S> {
...
+                                // We have an id, but it's not our response. Notify the thread and
+                                // move on
+                                trace!(
+                                    "Reader thread {}, received response for {}",
+                                    req.id,
+                                    resp_id
+                                );
+
+                                let mut map = self.waiting_map.write().unwrap();
+                                if map.get(&resp_id).is_some() {
```
Comment (RCasatta) [inline src/client.rs:343]: if let Some(sender) = map.remove(&resp_id) {
 sender.send(resp);
}

File: src/client.rs [hunk 2]
```diff
@@ -250,21 +262,119 @@ impl<S: Read + Write> Client<S> {
...
+                                    "Reader thread {} received a response for its request",
+                                    req.id
+                                );
+
+                                // Remove ourselves from the "waiting map"
+                                let mut map = self.waiting_map.write().unwrap();
+                                map.remove(&req.id);
+
+                                // If the map is not empty, we select a random thread to become the
+                                // new reader thread. We close its channel to "wake it up"
```
Comment (RCasatta) [inline src/client.rs:320]: maybe clearer to use an explicit enum as the value sent across the channel to receive data or be woke up, something like:
```
enum ValOrWake {
 Val(Value),
WakeUp,
}
```

File: src/client.rs (hunk 1 above)
Comment (afilini) [reply src/client.rs:343]: It was like you proposed initially, but then I realized I was removing something from the map while i was also borrowing it from the map and I decided to change it.

Maybe it's actually possible to get it to work, I'll try

File: src/client.rs (hunk 2 above)
Comment (afilini) [reply src/client.rs:320]: yeah that's a good idea!

**Comment by RCasatta** [comment]
ack 335fd49eb52f23e54b37b41761dc22f2f8114f8e

**Pull request merged**